File: Session_10_StringMatching/multi_language_implementations/kmp.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func computeLPSArray(pat string, M int, lps []int) {
	lenPrev := 0
	lps[0] = 0
	i := 1

	for i < M {
		if pat[i] == pat[lenPrev] {
			lenPrev++
			lps[i] = lenPrev
			i++
		} else {
			if lenPrev != 0 {
				lenPrev = lps[lenPrev-1]
			} else {
				lps[i] = 0
				i++
			}
		}
	}
}

func runKMP(pat, txt string) {
	M := len(pat)
	N := len(txt)
	lps := make([]int, M)

	computeLPSArray(pat, M, lps)

	fmt.Printf("Precomputed LPS Table for pattern \"%s\":\n", pat)
	fmt.Print("  Char: ")
	for i := 0; i < M; i++ {
		fmt.Printf("%c ", pat[i])
	}
	fmt.Println()
	fmt.Print("  LPS:  ")
	for i := 0; i < M; i++ {
		fmt.Printf("%d ", lps[i])
	}
	fmt.Println()
	fmt.Println()

	i := 0
	j := 0

	fmt.Println("KMP Search steps:")
	for (N - i) >= (M - j) {
		if pat[j] == txt[i] {
			i++
			j++
		}

		if j == M {
			fmt.Printf("  Pattern found at index: %d\n", i-j)
			j = lps[j-1]
		} else if i < N && pat[j] != txt[i] {
			if j != 0 {
				fmt.Printf("  Mismatch at txt[%d] ('%c') and pat[%d] ('%c'). Sliding to pat[%d]\n",
					i, txt[i], j, pat[j], lps[j-1])
				j = lps[j-1]
			} else {
				i++
			}
		}
	}
}

func main() {
	txt := "ABABDABACDABABCABAB"
	pat := "ABABCABAB"

	fmt.Println("=== Knuth-Morris-Pratt (KMP) Search ===")
	fmt.Printf("Text: \"%s\"\n", txt)
	fmt.Printf("Pattern: \"%s\"\n", pat)
	fmt.Println()

	runKMP(pat, txt)

	fmt.Println()
	fmt.Print("Press Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}

File: Session_10_StringMatching/multi_language_implementations/kmp.rs
use std::io::{self, BufRead};

fn compute_lps_array(pat: &str, m: usize, lps: &mut [usize]) {
    let pat_bytes = pat.as_bytes();
    let mut len_prev = 0;
    lps[0] = 0;
    let mut i = 1;

    while i < m {
        if pat_bytes[i] == pat_bytes[len_prev] {
            len_prev += 1;
            lps[i] = len_prev;
            i += 1;
        } else {
            if len_prev != 0 {
                len_prev = lps[len_prev - 1];
            } else {
                lps[i] = 0;
                i += 1;
            }
        }
    }
}

fn run_kmp(pat: &str, txt: &str) {
    let pat_bytes = pat.as_bytes();
    let txt_bytes = txt.as_bytes();
    let m = pat_bytes.len();
    let n = txt_bytes.len();
    
    let mut lps = vec![0; m];
    compute_lps_array(pat, m, &mut lps);

    println!("Precomputed LPS Table for pattern \"{}\":", pat);
    print!("  Char: ");
    for i in 0..m {
        print!("{} ", pat_bytes[i] as char);
    }
    println!();
    print!("  LPS:  ");
    for i in 0..m {
        print!("{} ", lps[i]);
    }
    println!("\n");

    let mut i = 0;
    let mut j = 0;

    println!("KMP Search steps:");
    while (n - i) >= (m - j) {
        if pat_bytes[j] == txt_bytes[i] {
            i += 1;
            j += 1;
        }

        if j == m {
            println!("  Pattern found at index: {}", i - j);
            j = lps[j - 1];
        } else if i < n && pat_bytes[j] != txt_bytes[i] {
            if j != 0 {
                println!(
                    "  Mismatch at txt[{}] ('{}') and pat[{}] ('{}'). Sliding to pat[{}]",
                    i, txt_bytes[i] as char, j, pat_bytes[j] as char, lps[j - 1]
                );
                j = lps[j - 1];
            } else {
                i += 1;
            }
        }
    }
}

fn main() {
    let txt = "ABABDABACDABABCABAB";
    let pat = "ABABCABAB";

    println!("=== Knuth-Morris-Pratt (KMP) Search ===");
    println!("Text: \"{}\"", txt);
    println!("Pattern: \"{}\"\n", pat);

    run_kmp(pat, txt);

    println!();
    println!("Press Enter to exit...");
    let mut iterator = io::stdin().lock().lines();
    let _ = iterator.next();
}
